amqp-client/src/main/java/zipkin2/reporter/amqp/RabbitMQSender.java

```diff
@@ -0,0 +1,246 @@
+      return this;
+    }
+
+    /** The AMQP user name to use when connecting to the broker. Defaults to "guest" */
+    public Builder username(String username) {
+      connectionFactory.setUsername(username);
+      return this;
+    }
+
+    /** The password to use when connecting to the broker. Defaults to "guest" */
+    public Builder password(String password) {
+      connectionFactory.setPassword(password);
+      return this;
+    }
+
+    public final RabbitMQSender build() {
+      return new AutoValue_RabbitMQSender(
+          encoding,
+          100000, // arbitrary to match kafka, messages theoretically can be 2GiB
+          addresses,
+          queue,
+          connectionFactory.clone(),
+          BytesMessageEncoder.forEncoding(encoding)
+      );
+    }
+  }
+
+  public final Builder toBuilder() {
+    return new Builder()
+        .connectionFactory(connectionFactory().clone())
+        .addresses(addresses())
+        .queue(queue())
+        .encoding(encoding());
+  }
+
+  abstract List<Address> addresses();
+
+  abstract String queue();
+
+  abstract ConnectionFactory connectionFactory();
+
+  abstract BytesMessageEncoder encoder();
+
+  /** get and close are typically called from different threads */
+  volatile boolean provisioned, closeCalled;
+
+  @Override public int messageSizeInBytes(List<byte[]> encodedSpans) {
+    return encoding().listSizeInBytes(encodedSpans);
+  }
+
+  /**
+   * This sends all of the spans as a single message.
+   *
+   * <p>NOTE: this blocks until the metadata server is available.
+   */
+  @Override public Call<Void> sendSpans(List<byte[]> encodedSpans) {
+    if (closeCalled) throw new IllegalStateException("closed");
+    byte[] message = encoder().encode(encodedSpans);
+    return new RabbitMQCall(message);
+  }
+
+  /** Ensures there are no connection issues. */
+  @Override public CheckResult check() {
+    try {
+      get().getHeartbeat();
+      return CheckResult.OK;
+    } catch (RuntimeException e) {
+      return CheckResult.failed(e);
+    }
+  }
+
+  @Memoized Connection get() {
+    Connection result;
+    try {
+      result = connectionFactory().newConnection(addresses());
+    } catch (IOException | TimeoutException e) {
+      throw new IllegalStateException("Unable to establish connection to RabbitMQ server", e);
+    }
+    provisioned = true;
+    return result;
+  }
+
+  @Override public synchronized void close() throws IOException {
+    if (closeCalled) return;
+    if (provisioned) get().close();
+    closeCalled = true;
+  }
+
+  class RabbitMQCall extends BaseCall<Void> { // RabbitMQFuture is not cancelable
+    private final byte[] message;
+
+    RabbitMQCall(byte[] message) {
+      this.message = message;
+    }
+
+    @Override protected Void doExecute() throws IOException {
+      publish();
+      return null;
+    }
+
+    void publish() throws IOException {
+      Channel channel = get().createChannel();
```

Comment:
I think this is the simplest way to make this sender thread safe, but it does add two additional network roundtrips for every send (one to create the channel, one to close the channel). For better publishing performance, each thread should reuse a channel or a channel cache could be utilized. I doubt whether this would make a significant difference in the "normal" case where we expect sending to only happen every second, but under more pressure, it might matter.